src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/DungeonNpcProfitOverlay.java

```diff
@@ -117,23 +119,42 @@ public void onButtonExclusionZones(ButtonExclusionZoneEvent event) {
 	@SubscribeEvent
 	public void onDrawBackground(GuiScreenEvent.BackgroundDrawnEvent event) {
 		if (!NotEnoughUpdates.INSTANCE.config.dungeons.croesusProfitOverlay || !(event.gui instanceof GuiChest)) {
-			chestProfits = null;
+			chestProfits.clear();
 			previousSlots = null;
 			return;
 		}
 
 		String lastOpenChestName = SBInfo.getInstance().lastOpenChestName;
 		Matcher matcher = chestNamePattern.matcher(lastOpenChestName);
 		if (!matcher.matches()) {
-			chestProfits = null;
+			chestProfits.clear();
 			previousSlots = null;
 			return;
 		}
 		GuiChest guiChest = (GuiChest) event.gui;
 		List<Slot> slots = guiChest.inventorySlots.inventorySlots;
 
-		if (chestProfits == null || chestProfits.isEmpty() || !slots.equals(previousSlots)) {
-			updateDungeonChests(slots);
+		if (chestProfits.isEmpty() || !slots.equals(previousSlots)) {
+			chestProfits.clear();
+			updateDungeonChests(slots).thenAccept(chests -> {
+				for (DungeonChest chest : chests) {
+					chest.calculateProfitAndBuildLore();
+					chestProfits.removeIf(c -> c.name.equals(chest.name) && c.items.size() <= chest.items.size());
```

Comment:
I still dont like that you are modifying the list instead of performing all the calculations off thread and then using a single assignment to transfer the data over. I really think that could help remove a lot of those synchronization issues.